prog/dftb+/lib_timedep/dynamics.F90

```diff
@@ -3597,11 +3381,603 @@ subroutine getBondPopulAndEnergy(this, bondWork, lastBondPopul, rhoPrim, ham0, o
+          call propagateRho(this, this%rhoOld(:,:,iKS), this%rho(:,:,iKS),&
+              & this%H1(:,:,iKS), this%Sinv(:,:,iKS), this%dt)
+        else
+          call propagateRho(this, this%rhoOld(:,:,iKS), this%rho(:,:,iKS),&
+              & this%H1(:,:,iKS), this%Sinv(:,:,iKS), 2.0_dp * this%dt)
+        end if
+      else
+        call propagateRhoRealH(this, this%rhoOld(:,:,iKS), this%rho(:,:,iKS),&
+            & this%H1(:,:,iKS), this%Sinv(:,:,iKS), 2.0_dp * this%dt)
+      end if
+    end do
+
+    if (mod(iStep, 2) == 1) then
+      this%rho => this%trho
+      this%rhoOld => this%trhoOld
+    else
+      this%rho => this%trhoOld
+      this%rhoOld => this%trho
+    end if
+
+    if ((this%tPopulations) .and. (mod(iStep, this%writeFreq) == 0)) then
+      do iKS = 1, this%parallelKS%nLocalKS
+        call getTDPopulations(this, this%occ, this%rho, this%Eiginv, this%EiginvAdj, this%populDat, this%time, iKS)
+      end do
+    end if
+
+  end subroutine doTdStep
+
+  !> Handles deallocation, closing outputs and autotest writing
+  subroutine finalizeDynamics(this, coord, tWriteAutotest, taggedWriter)
+
+    !> ElecDynamics instance
+    type(TElecDynamics), intent(inout) :: this
+
+    !> atomic coordinates
+    real(dp), allocatable, intent(inout) :: coord(:,:)
+
+    !> Should autotest data be written?
+    logical, intent(in) :: tWriteAutotest
+
+    !> Tagged writer object
+    type(TTaggedWriter), intent(inout) :: taggedWriter
+
+    if (tWriteAutotest) then
+      call writeTDAutotest(this, this%dipole, this%energy, this%deltaQ, coord, this%totalForce,&
+          & this%occ, this%lastBondPopul, taggedWriter)
+    end if
+
+    call closeTDOutputs(this, this%dipoleDat, this%qDat, this%energyDat, this%populDat,&
+        & this%forceDat, this%coorDat, this%fdBondPopul, this%fdBondEnergy)
+
+    deallocate(this%Ssqr)
+    deallocate(this%Sinv)
+    deallocate(this%H1)
+    deallocate(this%RdotSprime)
+    deallocate(this%qq)
+    deallocate(this%deltaQ)
+    deallocate(this%dipole)
+    deallocate(this%chargePerShell)
+    deallocate(this%occ)
+    deallocate(this%totalForce)
+    deallocate(this%trho)
+    deallocate(this%trhoOld)
+
+    deallocate(this%rhoPrim)
+    deallocate(this%ErhoPrim)
+    if (allocated(this%H1LC)) then
+      deallocate(this%H1LC)
+    end if
+    if (allocated(this%deltaRho)) then
+      deallocate(this%deltaRho)
+    end if
+    if (allocated(this%qBlock)) then
+      deallocate(this%qBlock)
+    end if
+    if (allocated(this%qNetAtom)) then
+      deallocate(this%qNetAtom)
+    end if
+    if (this%tPopulations) then
+      deallocate(this%Eiginv)
+      deallocate(this%EiginvAdj)
+    end if
+    if (this%tBondE .or. this%tBondP) then
+      deallocate(this%bondWork)
+    end if
+
+    if (this%tIons) then
+      if (allocated(this%polDirs)) then
+        if (size(this%polDirs) <  (this%nDynamicsInit + 1)) then
+          deallocate(this%pMDIntegrator)
+        end if
+      end if
```

Comment:
I am wondering, whether all those deallocations are necessary? (Usually, it is difficult to keep so many deallocates consistent with the correspoinding allocations.) Is it possible, to call `doDynamics` more than once? If not, we could leave them away and trust, that the caller dealocates `TElecDynamics` if it wants to make sure that all resources are freed.

Reply:
Thanks for the observation Bálint. In fact `doDynamics` can be called more than once (this is what runDynamics does), so it's not possible to deallocate `TElecDynamics`. The number of allocated variables in this module is large but still managable, so I could check manually that all which are allocated, are eventually deallocated. Would that work for the moment? The alternative would be to run valgrind and find memory leaks but that could be painful.